ambry-router/src/main/java/com.github.ambry.router/GetBlobInfoOperation.java

```diff
@@ -60,19 +60,31 @@
    * @param blobIdStr the blob id associated with the operation in string form.
    * @param futureResult the future that will contain the result of the operation.
    * @param callback the callback that is to be called when the operation completes.
+   * @param operationCompleteCallback the {@link OperationCompleteCallback} to use to complete operations.
    * @param time the Time instance to use.
    * @throws RouterException if there is an error with any of the parameters, such as an invalid blob id.
    */
   GetBlobInfoOperation(RouterConfig routerConfig, ClusterMap clusterMap, ResponseHandler responseHandler,
-      String blobIdStr, FutureResult<BlobInfo> futureResult, Callback<BlobInfo> callback, Time time)
+      String blobIdStr, FutureResult<BlobInfo> futureResult, Callback<BlobInfo> callback,
+      OperationCompleteCallback operationCompleteCallback, Time time)
       throws RouterException {
     super(routerConfig, clusterMap, responseHandler, blobIdStr, futureResult, callback, time);
+    this.operationCompleteCallback = operationCompleteCallback;
     operationTracker = new SimpleOperationTracker(routerConfig.routerDatacenterName, blobId.getPartition(),
         routerConfig.routerGetCrossDcEnabled, routerConfig.routerGetSuccessTarget,
         routerConfig.routerGetRequestParallelism);
   }
 
   /**
+   * {@inheritDoc}
+   */
+  @Override
+  void abort(Exception abortCause) {
+    operationCompleteCallback.completeOperation(operationFuture, operationCallback, null, abortCause);
+    operationCompleted = true;
```

Comment:
Perhaps not. But it made sense to do this in case code elsewhere assumed that if `operationCompleted` is true, then the operation was certainly completed. For example, if, for whatever reason, somewhere we chose to do this:

```java
if (operationCompleted) {
  // clear the state of the operation
  operationFuture = null;
  operationCallback = null;
}
```
Though this does not arise today. Was also trying to be consistent with the way we do this for chunks, where we mark the state as ready or complete or free only after all the work is done (which matters for the chunks).